### ISSUE TITLE
Large negative cdf for values near tau

### ISSUE THREAD

**Issue by itsdfish** [itsdfish/SequentialSamplingModels.jl#23]
@kiante-fernandez, I noticed that `cdf` returns a large negative value for values near tau. Do you know if this is an expected edge case or perhaps a bug? I suspect it might be an edge case, but I don't know for certain. 
```
using SequentialSamplingModels
dist = DDM(ν=2.0, α = 1.5, z = .5, τ = .30) 
cdf(dist, 1, 0.305)
```

**Comment by kiante-fernandez** [comment]
I was looking at the supplementary material for the [Blurton et al., 2012 paper](https://ars.els-cdn.com/content/image/1-s2.0-S0022249612001009-mmc1.pdf).

I went ahead and tried the source implementation (in R) to check for edge case. It didn't occur, but it made me think that our implementation might benefit from defining a new variable that represents "t - τ" instead of subtracting non-decision time each time in the implementation. Maybe that could help us identify the bug? 

```R
# Calculates exp(a) * pnorm(b) using an approximation by Kiani et al. (2008)
exp_pnorm = function(a, b)
{
  r = exp(a) * pnorm(b)
  d = is.nan(r) & b < -5.5
  r[d] = 1/sqrt(2) * exp(a - b[d]*b[d]/2) * (0.5641882/b[d]/b[d]/b[d] - 1/b[d]/sqrt(pi))
  r
}
# Number of terms required for large time representation
K_large = function(t, v, a, w, epsilon)
{
  sqrtL1 = sqrt(1/t) * a/pi
  sqrtL2 = sqrt(pmax(1, -2/t*a*a/pi/pi * (log(epsilon*pi*t/2 * (v*v + pi*pi/a/a)) + v*a*w + v*v*t/2)))
  ceiling(pmax(sqrtL1, sqrtL2))
}
# Number of terms required for small time representation
K_small = function(t, v, a, w, epsilon)
{
  if(abs(v) < sqrt(.Machine$double.eps)) # zero drift case
    return(ceiling(pmax(0, w/2 - sqrt(t)/2/a * qnorm(pmax(0, pmin(1, epsilon/(2-2*w)))))))
  if(v > 0) # positive drift
    return(K_small(t, -v, a, w, exp(-2*a*w*v)*epsilon))
  S2 = w - 1 + 1/2/v/a * log(epsilon/2 * (1-exp(2*v*a)))
  S3 = (0.535 * sqrt(2*t) + v*t + a*w)/2/a
  S4 = w/2 - sqrt(t)/2/a * qnorm(pmax(0, pmin(1, epsilon * a / 0.3 / sqrt(2*pi*t) * exp(v*v*t/2 + v*a*w))))
  ceiling(pmax(S2, S3, S4, 0))
}
# Probability for absorption at upper barrier
Pu = function(v, a, w)
{
  e = exp(-2 * v * a * (1-w))
  if(e == Inf)
    return(1)
  if(abs(e - 1) < sqrt(.Machine$double.eps)) # drift near zero or w near 1
    return(1 - w)
  (1 - e) / (exp(2*v*a*w) - e) # standard case
}
# Large time representation of lower subdistribution
Fl_lower = function(t, v, a, w, K)
{
  F = numeric(length(t))
  for(k in K:1)
    F = F - k / (v*v + k*k*pi*pi/a/a) * exp(-v*a*w - 1/2*v*v*t - 1/2*k*k*pi*pi/a/a*t) * sin(pi*k*w)
  Pu(v, a, w) + 2*pi/a/a * F
} 
#Online supplementary material for Blurton et al.
# Small time representation of the upper subdistribution
Fs_lower = function(t, v, a, w, K)
{
  if(abs(v) < sqrt(.Machine$double.eps)) # zero drift case
    return(Fs0_lower(t, a, w, K))
  S1 = S2 = numeric(length(t))
  sqt = sqrt(t)
  for(k in K:1)
  {
    S1 = S1 + exp_pnorm(2*v*a*k, -sign(v)*(2*a*k+a*w+v*t)/sqt) -
      exp_pnorm(-2*v*a*k - 2*v*a*w, sign(v)*(2*a*k+a*w-v*t)/sqt)
    S2 = S2 + exp_pnorm(-2*v*a*k, sign(v)*(2*a*k-a*w-v*t)/sqt) -
      exp_pnorm(2*v*a*k - 2*v*a*w, -sign(v)*(2*a*k-a*w+v*t)/sqt)
  }
  Pu(v, a, w) + sign(v) * ((pnorm(-sign(v) * (a*w+v*t)/sqt) -
                              exp_pnorm(-2*v*a*w, sign(v) * (a*w-v*t)/sqt)) + S1 + S2)
}
# Zero drift version
Fs0_lower = function(t, a, w, K)
{
  F = numeric(length(t))
  for(k in K:0)
    F = F - pnorm((-2*k - 2 + w)*a/sqrt(t)) + pnorm((-2*k - w)*a/sqrt(t))
  2*F
}
# Lower subdistribution
F_lower = function(t, v, a, w, sigma2, epsilon)
{
  a = a / sqrt(sigma2)
  v = v / sqrt(sigma2)
  K_l = K_large(t, v, a, w, epsilon)
  K_s = K_small(t, v, a, w, epsilon)
  F = numeric(length(t))
  i = (K_l < 10*K_s)
  if(any(i)) F[i] = Fl_lower(t[i], v, a, w, max(K_l[i]))
  if(any(!i)) F[!i] = Fs_lower(t[!i], v, a, w, max(K_s[!i]))
  F
}
# Upper subdistribution
F_upper = function(t, v, a, w, sigma2, epsilon)
{
  F_lower(t, -v, a, 1-w, sigma2, epsilon)
} 

F_lower(t = 0.305, v = 2.0, a = 1.5, w = .5, sigma2 = 1,  epsilon = 1.0e-12)

```

**Comment by itsdfish** [comment]
Thanks @kiante-fernandez. This is helpful. I think the best approach is to define unit tests which target the helper functions above. In the process of setting up the tests, I found one issue. The following test disagrees with the R code (assuming it is the ground truth):

```
test_val2 = _K_large(DDM(;ν=1.5, α=.50, z=.25, τ=.50), .501; ϵ = 1e-10)
 @test test_val2 ≈ 36
```
which is based on `K_large(.001, 1.5, .5, .25, 1e-10)`

The Julia code returns 37. I have not had time yet to diagnose the problem, but it could potentially be related to the negative value.  I'll continue adding unit tests to the `ddm_cdf` branch as time permits.

**Comment by itsdfish** [comment]
I'm also not finding `F_upper` in the Julia code. I thought maybe it is integrated into another function, but I'm not seeing it.

**Comment by itsdfish** [comment]
I found three errors in the function `_Fs_lower`. `cdf` tests now work with t near tau. Another issue is that the for loop in this function does not work for K=0. I"m not sure if that is ever a value that is used.

 @kiante-fernandez, would you like to look into the issue with `_K_large`?

**Comment by kiante-fernandez** [comment]
> I'm also not finding `F_upper` and `F_lower` in the Julia code. I thought maybe they are integrated into another function, but I'm not seeing them.

They are the `cdf` in our case.

**Comment by kiante-fernandez** [comment]
> I found three errors in the function `_Fs_lower`. `cdf` tests now work with t near tau. Another issue is that the for loop in this function does not work for K=0. I"m not sure if that is ever a value that is used.
> 
> @kiante-fernandez, would you like to look into the issue with `_K_large`?

Yep I can take a look today!

**Comment by kiante-fernandez** [comment]
> @kiante-fernandez, I wanted to follow up in case you didn't see my chat message at ICCM. What a nice surprise! I won't be going to the in person event due to a knee injury unfortunately. It was nice running into you.

Ah, I just realized I missed it. My internet connection was quite bad. It was a pleasant surprise to run into you. It's nice to know we are involved in similar circles :)

**Comment by kiante-fernandez** [comment]
> I found three errors in the function `_Fs_lower`. `cdf` tests now work with t near tau. Another issue is that the for loop in this function does not work for K=0. I"m not sure if that is ever a value that is used.
> 
> @kiante-fernandez, would you like to look into the issue with `_K_large`?

```R
F_lower(t = 0.005, v = 2.0, a = 1.5, w = .5, sigma2 = 1,  epsilon = 1.0e-12)
```
will also return a negative value. looks like a special case. 

I ran the test for the `_K_large` function. The only difference in results between the R and Julia `_K_large` functions was the misplacement of a (). I will pass the test with the next pull request. I rewrote it to more explicitly follow the R implementation.

I also found some bugs in `_Fs_lower`. Also fixed. 

I will put the PR once I do some final checks on the rest of the functions.

**Comment by itsdfish** [comment]
Hi Kiante,

Thanks for looking into that. On Sunday,  I did fix some bugs in `_Fs_lower` and write some tests on branch `ddm_cdf` (see above). I went ahead and merged those changes into master. Can you pull those changes and resolve with your PR?

**Comment by kiante-fernandez** [comment]
I think we can close this now.

**Comment by itsdfish** [comment]
Yup. Thanks for the reminder. 

By the way, I think you can do something like closes #issue_number, to automatically close the issue once the PR is merged.
